[IR-2053] Unit Tests: Core All ECS Hooks/Reactors

heysokam opened the pull request:
## Summary
- [x] useQuery
- [x] ~~useAllComponents~~
- [x] useComponent
- [x] useOptionalComponent
- [x] UUIDComponent

## Subtasks Checklist

## Breaking Changes

## References

## QA Steps

heysokam committed:
fix: Typo in `SystemFunctions.test.ts`

heysokam committed:
fix: Typo in `ComponentFunctions.test.ts`

heysokam committed:
Merge branch 'dev' into IR-2053-unit-test-core-all-ecs-hooks-reactors

heysokam committed:
tst: Skeleton for the `useComponent` test cases

heysokam committed:
Merge branch 'dev' into IR-2053-unit-test-core-all-ecs-hooks-reactors

heysokam committed:
chg: `useAllComponents` initializes the result with `getAllComponents`

heysokam committed:
tst: UnitTests for `ComponentFunctions.ts` hooks

heysokam committed:
tst: UnitTests for `UUIDComponent.ts` (phase 0)

heysokam committed:
Merge branch 'dev' into IR-2053-unit-test-core-all-ecs-hooks-reactors

heysokam committed:
tst: UnitTests for `UUIDComponent.ts` (phase 1)

heysokam committed:
tst: Small change to the `useEntityByUUID` hook test

DanielBelmes commented on packages/ecs/src/UUIDComponent.test.tsx:
```diff
@@ -0,0 +1,249 @@
+      assert.equal(component1, TestUUID, 'The UUID value did not get set correctly')
+    })
+
+    it("shouldn't change the data when set multiple times with the same data", () => {
```
What happends if we set UUIDComponent with a different UUID? Do we throw an error if we try to change it?

heysokam replied:
Nothing. It returns early

DanielBelmes commented on packages/ecs/src/UUIDComponent.test.tsx:
```diff
@@ -0,0 +1,249 @@
+    it('should return the correct entity', () => {
+      setComponent(entity1, UUIDComponent, TestUUID)
+      const testEntity = UUIDComponent.getEntityByUUID(TestUUID)
+      assert.equal(testEntity, UUIDComponent.entitiesByUUIDState[TestUUID].value)
```
you can just pass entity1 into this condition as testEntity should be the same type and value as entity1.

heysokam replied:
The idea of that `entitiesByUUIDState` check is to confirm that the underlying data is the same as the expected value, independent of what the function is doing internally
Added the comment as an extra check with c37140192ebfb47c3732e8c091ba64c751bd3638

DanielBelmes replied:
oh yea reading this back two weeks later I was dumb.

heysokam committed:
tst: Extra check for the `getEntityByUUID` test

heysokam committed:
tst: Small formatting change to one of the assertion messages

heysokam committed:
tst: `npm run test` vscode runner for the `ecs` package

heysokam committed:
tst: UnitTests for `QueryFunctions.ts`

heysokam committed:
fix: Error in `useEntityByUUID` tests

HexaField commented on packages/ecs/src/QueryFunctions.test.tsx:
```diff
@@ -0,0 +1,181 @@
+  })
+
+  /**
+   *  @todo : Why does `query()` still work after calling `removeQuery(query)`
```
This is a design decision of bitecs. Queries are defined at global scope, but only initialized lazily when they are first called. Removing a query simply releases it's instance inside bitecs until it is called again. All queries are automatically cleaned up when the bitecs world is destroyed (which happens when the hyper store / engine is destroyed)

HexaField commented on packages/ecs/src/QueryFunctions.test.tsx:
```diff
@@ -0,0 +1,181 @@
+  })
+
+  /**
+   *  @todo : Why does `query()` still work after calling `removeQuery(query)`
```
we can remove this commented out test

heysokam replied:
Removed with 2cf1110c793eee66fbb7e8d3f634bdd31e8c9bd7

HexaField commented on packages/ecs/src/QueryFunctions.test.tsx:
```diff
@@ -0,0 +1,181 @@
+    // Define the Reactor that will run the tested hook
+    const Reactor = () => {
+      const data = useQuery([component])
+      console.log('render', data)
```
remove console logs

heysokam replied:
Applied on 2cf1110c793eee66fbb7e8d3f634bdd31e8c9bd7

HexaField commented on packages/ecs/src/QueryFunctions.test.tsx:
```diff
@@ -0,0 +1,181 @@
+      const tag = <Reactor />
+      const { rerender, unmount } = render(tag)
+      await act(() => rerender(tag))
+      assertArrayNotEqual(
```
i'm not a fan of asserting against some arbitrary value. we should just be asserting the data we expect. this is functionally equivalent as if `result` is undefined

heysokam replied:
Yes, but you cannot check for undefined coming from the reactor if you don't initialize with some dummy data before doing so.

With this check, you can test if the reactor touched the result, and also check if the reactor applied an `undefined` value to the result (like with optional values, etc)

That's the reasoning behind it.

HexaField replied:
we should be asserting both that the value is undefined and that the useEffect has run. something like a counter increment along with the result data would suffice

heysokam replied:
Changed on 0320072
@HexaField Let me know if the `counter` variable checks are correct :writing_hand:

HexaField commented on packages/ecs/src/QueryFunctions.test.tsx:
```diff
@@ -0,0 +1,181 @@
+import { createEntity, removeEntity } from './EntityFunctions'
+import { Query, defineQuery, useQuery } from './QueryFunctions'
+
+function assertArrayEqual<T>(A: Array<T>, B: Array<T>, err = 'Arrays are not equal') {
```
are these helpers necessary? does `assert.deepEqual` not work for just two arrays?

heysokam replied:
I don't know, honestly.
I tried with `equal` and it was failing because their references were different even if their data was equal.
I will try deepEqual :writing_hand:

heysokam replied:
Well, apparently I did try it before.
It doesn't work with `assert.deepEqual` directly on the arrays without iterating :shrug:

heysokam committed:
Merge branch 'dev' into IR-2053-unit-test-core-all-ecs-hooks-reactors

heysokam committed:
tst: Apply review suggestions

heysokam committed:
Merge branch 'dev' into IR-2053-unit-test-core-all-ecs-hooks-reactors

heysokam committed:
tst: Replace instances of `InitialValue` with `undefined & counter`

As per review:
- Leftover console.log calls removed
- InitialValue checks swapped with `undefined` and a `counter` variable

For Types readability:
- Change type casts to use a `ResultType` that unions undefined with the expected returned type

heysokam committed:
Merge branch 'dev' into IR-2053-unit-test-core-all-ecs-hooks-reactors

heysokam committed:
fix: Typescript type conflicts

heysokam committed:
Merge branch 'dev' into IR-2053-unit-test-core-all-ecs-hooks-reactors

heysokam committed:
tst: Remove `useAllComponents` tests. Hook removed at #a1a6682a

heysokam committed:
Merge branch 'dev' into IR-2053-unit-test-core-all-ecs-hooks-reactors

DanielBelmes commented:
Has conflicts @heysokam. Though I'm not sure how they got there since this PR creates the file

DanielBelmes reviewed (approved):
Preliminary approval. Still pending integration with @speigg tests he wrote for immedieteUseEffect.

heysokam committed:
tst: Remove `QueryFunctions.test.tsx` for merge conflict resolution

heysokam committed:
Merge branch 'dev' into IR-2053-unit-test-core-all-ecs-hooks-reactors

heysokam committed:
tst: Add UnitTests to `QueryFunctions.test.tsx` after merge conflict

heysokam commented:
> Preliminary approval. Still pending integration with @speigg tests he wrote for immedieteUseEffect.

@DanielBelmes Should be good to go now :writing_hand:
